### PR TITLE
make some files cppcheck clean

### PR DESCRIPTION
**Pull request by heckflosse**
@Desmis Jacques, can you please test my changes to color.cc?

### CODE REVIEW

File: rtengine/FTblockDN.cc [hunk 1]
```diff
@@ -2141,18 +2141,18 @@ float ImProcFunctions::MadRgb(const float * DataList, const int datalen)
...
+        ++lmedian;
     }
 
-    int count_ = count - histo[median - 1];
+    int count_ = count - histo[lmedian - 1];
 
     // interpolate
     delete[] histo;
-    return (((median - 1) + (datalen / 2 - count_) / (static_cast<float>(count - count_))) / 0.6745f);
+    return ((lmedian - 1) + (datalen / 2 - count_) / (static_cast<float>(count - count_))) / 0.6745f;
```
Comment (Floessie) [inline rtengine/FTblockDN.cc:2155]: Shouldn't this be factored out (anon namespace)? It's mostly a duplicate of the above...

File: rtengine/ashift_lsd.c [hunk 2]
```diff
@@ -1770,10 +1770,10 @@ static void region_grow( int x, int y, image_double angles, struct point * reg,
   int xx,yy,i;
 
   /* check parameters */
-  if( x < 0 || y < 0 || x >= (int) angles->xsize || y >= (int) angles->ysize )
-    error("region_grow: (x,y) out of the image.");
   if( angles == NULL || angles->data == NULL )
     error("region_grow: invalid image 'angles'.");
+  if( x < 0 || y < 0 || x >= (int) angles->xsize || y >= (int) angles->ysize )
```
Comment (Floessie) [inline rtengine/ashift_lsd.c:1775]: Good catch! :+1:

File: rtengine/color.cc [hunk 3]
```diff
@@ -1189,145 +1189,109 @@ void Color::computeBWMixerConstants (const Glib::ustring &setting, const Glib::u
...
+            const float pgM = complement ? (-0.492f * mixerPurple + 16.236f) / 100.f : 0.f;
 
             mixerRed   += prM;
             mixerGreen += pgM;
             mixerBlue  += pbM;
             koymcp += (prM + pgM + pbM);
         }
 
-        if(mixerCyan != 33) {
+        if(mixerCyan != 33.f) {
```
Comment (Floessie) [inline rtengine/color.cc]: Space.

File: rtengine/color.cc [hunk 4]
```diff
@@ -1496,100 +1460,78 @@ void Color::interpolateRGBColor (float realL, float iplow, float iphigh, int alg
...
-                b_1 = b_1 + (b_2 - b_1) * calan * balance;
-            } else if(metchrom == 2) {
-                a_1 = a_1 + (a_2 - a_1) * calan * balance;
-                b_1 = b_1 + (b_2 - b_1) * calby * balance;
+            } else if (metchrom == 1) {
+                a_1 = a_1 + (a_2 - a_1) * chromat * balance;
+                b_1 = b_1 + (b_2 - b_1) * chromat * balance;
+            } else if (metchrom == 2) {
+                a_1 = a_1 + (a_2 - a_1) * chromat * balance;
+                b_1 = b_1 + (b_2 - b_1) * luma * balance;
```
Comment (Floessie) [inline rtengine/color.cc:1522]: As always, having proper `enum`s would ease reading the code and would also enable `switch` without `default:` to warn about missing `case`s. :expressionless:

File: rtengine/color.cc [hunk 5]
```diff
@@ -1496,100 +1460,78 @@ void Color::interpolateRGBColor (float realL, float iplow, float iphigh, int alg
                                  const float xl, const float yl, const float zl, const float x2, const float y2, const float z2,
                                  const double xyz_rgb[3][3], const double rgb_xyz[3][3], float &ro, float &go, float &bo)
 {
-    float X1, Y1, Z1, X2, Y2, Z2, X, Y, Z, XL, YL, ZL;
-    float L1 = 0.f, L2, LL, a_1 = 0.f, b_1 = 0.f, a_2 = 0.f, b_2 = 0.f, a_L, b_L;
+    float L1 = 0.f, a_1 = 0.f, b_1 = 0.f, a_2 = 0.f, b_2 = 0.f, a_L, b_L;
```
Comment (Floessie) [inline rtengine/color.cc]: Is it worth it to omit the `a_L` and `b_L` initialization?

File: rtengine/color.cc [hunk 6]
```diff
@@ -1496,100 +1460,78 @@ void Color::interpolateRGBColor (float realL, float iplow, float iphigh, int alg
...
     Color::Lab2XYZ(L1, a_1, b_1, X, Y, Z);
-
     Color::xyz2rgb(X, Y, Z, ro, go, bo, rgb_xyz);// ro go bo in gamut
 }
 
 void Color::calcGamma (double pwr, double ts, int mode, GammaValues &gamma)
 {
     //from Dcraw (D.Coffin)
-    int i;
     double g[6], bnd[2] = {0., 0.};
```
Comment (Floessie) [inline rtengine/color.cc]: ` = {}`?

File: rtengine/color.cc [hunk 7]
```diff
@@ -1621,7 +1563,7 @@ void Color::calcGamma (double pwr, double ts, int mode, GammaValues &gamma)
         g[5] = 1. / (g[1] * SQR(g[3]) / 2. + 1. - g[2] - g[3] - g[2] * g[3] * (log(g[3]) - 1.)) - 1.;
     }
 
-    if (!mode--) {
+    if (!mode) {
```
Comment (Floessie) [inline rtengine/color.cc]: Why? If `mode != 0` there is no result?!

File: rtengine/color.cc [hunk 8]
```diff
@@ -2078,15 +2020,11 @@ void Color::gamutmap(float &X, float &Y, float &Z, const double p[3][3])
...
-    float scale = 100.0f / 100.1f; //reduction in normal zone
-    float scaleext = 1.0f; //reduction in transition zone
-    float deltaHH = 0.3f; //HH value transition : I have choice 0.3 radians
+
     float HH;
     bool doskin = false;
-
     //rough correspondence between h (JC) and H (lab) that has relatively little importance because transitions that blur the correspondence is not linear
-    if     ((float)h > 8.6f  && (float)h <= 74.f ) {
+    if ((float)h > 8.6f  && (float)h <= 74.f ) {
```
Comment (Floessie) [inline rtengine/color.cc]: I'm pretty sure the cast is implicit.

File: rtengine/color.cc [hunk 9]
```diff
@@ -2117,22 +2058,23 @@ void Color::skinred ( double J, double h, double sres, double Sp, float dred, fl
...
-        if(chromapro > 1.0) {
-            interm = (chromapro - 1.0f) * 100.0f;
+        float factorskinext;
+        if(chromapro > 1.f) {
+            const float interm = (chromapro - 1.0f) * 100.0f;
             factorskin = 1.0f + (interm * scale) / 100.0f;
             factorskinext = 1.0f + (interm * scaleext) / 100.0f;
         } else {
             factorskin = chromapro ;
             factorskinext = chromapro ;
```
Comment (Floessie) [inline rtengine/color.cc]: This needs to be run through a linter...

File: rtengine/color.cc [hunk 10]
```diff
@@ -2296,26 +2227,19 @@ void Color::AllMunsellLch(bool lumaMuns, float Lprov1, float Loldd, float HH, fl
...
-
-                        if(contin1 && contin2) {
-                            correctlum = correctlumprov2 - correctlumprov;
+                            if(contin1) {
+                                correctlum = correctionHueLum - correctlumprov;
+                            }
                         }
                     }
                 }
+                break;
```
Comment (Floessie) [inline rtengine/color.cc:2169]: Optimization?

File: rtengine/color.h [hunk 11]
```diff
@@ -1805,7 +1805,7 @@ class Color
     * @param Z Z channel input value and corrected output value [0 ; 65535]
     * @param p working profile
     */
-    static void gamutmap(float &X, float &Y, float &Z, const double p[3][3]);
+    static void gamutmap(float &X, float Y, float &Z, const double p[3][3]);
```
Comment (Floessie) [inline rtengine/color.h:1804]: You should adjust the doxygen description.

File: rtengine/imagefloat.h [hunk 12]
```diff
@@ -109,35 +109,35 @@ class Imagefloat final : public IImagefloat, public ImageIO
         } tmp;
 
         tmp.f = f;
-        int32_t sign     =  (tmp.i >> 16) & 0x00008000;
+        const int32_t lsign = (tmp.i >> 16) & 0x00008000;
```
Comment (Floessie) [inline rtengine/imagefloat.h:112]: `std::int32_t`

File: rtengine/imagefloat.h [hunk 13]
```diff
@@ -214,7 +214,7 @@ class Imagefloat final : public IImagefloat, public ImageIO
         exponent += (128 - 64);
         mantissa <<= 7;
         // Assemble sign, exponent and mantissa.
-        return (uint32_t) ((sign << 31) | (exponent << 23) | mantissa);
+        return (uint32_t) ((lsign << 31) | (exponent << 23) | mantissa);
```
Comment (Floessie) [inline rtengine/imagefloat.h:217]: Does this really have to be inline in the (or *this*) header?

**Review by Floessie — COMMENTED** [review]
Just a few comments.

**Comment by Desmis** [comment]
@heckflosse 

I tested with various change in BW, Color toning....
No differences in TIF

@Floessie 
i am still impressed :)

Jacques

File: rtengine/color.cc (hunk 7 above)
Comment (heckflosse) [reply rtengine/color.cc]: Fixed

File: rtengine/color.cc (hunk 6 above)
Comment (heckflosse) [reply rtengine/color.cc]: Fixed

File: rtengine/color.cc (hunk 3 above)
Comment (heckflosse) [reply rtengine/color.cc]: Fixed

File: rtengine/color.cc (hunk 8 above)
Comment (heckflosse) [reply rtengine/color.cc]: I removed this (unused) function.

File: rtengine/color.cc (hunk 9 above)
Comment (heckflosse) [reply rtengine/color.cc]: Function removed

File: rtengine/color.cc (hunk 10 above)
Comment (heckflosse) [reply rtengine/color.cc:2169]: Yes

File: rtengine/color.h (hunk 11 above)
Comment (heckflosse) [reply rtengine/color.h:1804]: Fixed

File: rtengine/imagefloat.h (hunk 12 above)
Comment (heckflosse) [reply rtengine/imagefloat.h:112]: We should refactor this stuff, as it's duplicated code. Have a look:
https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/dcraw.cc#L10709

File: rtengine/imagefloat.h (hunk 13 above)
Comment (heckflosse) [reply rtengine/imagefloat.h:217]: As above

File: rtengine/color.cc (hunk 5 above)
Comment (heckflosse) [reply rtengine/color.cc]: Fixed

**Comment by heckflosse** [comment]
@Desmis Jacques, 
here
https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/dirpyr_equalizer.cc#L272
cppcheck correctly detected that the condition `level == 5` can never be true because above there is the condition `level != 5`.

Is this a typo? How should I fix this?

**Comment by Desmis** [comment]
@heckflosse 

The good choice is in   "if(blurcb > 0.f && choice == 0  && level != 5) {" to suppress level != 5

But I disabled this function by setting "blurcb" to zero (and also for GUI , I suppress slider "blur"), because it brought artifacts as soon as the radius was not very small and only for low levels.
The same function in wavelets works  (local and main) fine with much more functionality, for all levels and for high radius.

If we compare "wavelet" with "cbdl", in all cases (whether local or main) wavelet is clearly superior to cbdl ... 
The only reasons to maintain "cbdl" are the habit and an apparent simplicity (wavelet seems complex ...)

jacques

**Pull request merged**